// nucleation-wasm/nucleation.js
let wasm;

function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_externrefs.set(idx, obj);
    return idx;
}

function _assertClass(instance, klass) {
    if (!(instance instanceof klass)) {
        throw new Error(`expected instance of ${klass.name}`);
    }
}

function getArrayF64FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getFloat64ArrayMemory0().subarray(ptr / 8, ptr / 8 + len);
}

let cachedFloat64ArrayMemory0 = null;
function getFloat64ArrayMemory0() {
    if (cachedFloat64ArrayMemory0 === null || cachedFloat64ArrayMemory0.byteLength === 0) {
        cachedFloat64ArrayMemory0 = new Float64Array(wasm.memory.buffer);
    }
    return cachedFloat64ArrayMemory0;
}

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}

let cachedUint32ArrayMemory0 = null;
function getUint32ArrayMemory0() {
    if (cachedUint32ArrayMemory0 === null || cachedUint32ArrayMemory0.byteLength === 0) {
        cachedUint32ArrayMemory0 = new Uint32Array(wasm.memory.buffer);
    }
    return cachedUint32ArrayMemory0;
}

let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

function passArray32ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 4, 4) >>> 0;
    getUint32ArrayMemory0().set(arg, ptr / 4);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArrayF64ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 8, 8) >>> 0;
    getFloat64ArrayMemory0().set(arg, ptr / 8);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = cachedTextEncoder.encodeInto(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
cachedTextDecoder.decode();
const MAX_SAFARI_DECODE_BYTES = 2146435072;
let numBytesDecoded = 0;
function decodeText(ptr, len) {
    numBytesDecoded += len;
    if (numBytesDecoded >= MAX_SAFARI_DECODE_BYTES) {
        cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
        cachedTextDecoder.decode();
        numBytesDecoded = len;
    }
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

const cachedTextEncoder = new TextEncoder();

if (!('encodeInto' in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function (arg, view) {
        const buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length
        };
    }
}

let WASM_VECTOR_LEN = 0;

const CompressionModelFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_compressionmodel_free(ptr >>> 0, 1));

const DetectorConfigFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_detectorconfig_free(ptr >>> 0, 1));

const NucleationDetectorFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_nucleationdetector_free(ptr >>> 0, 1));

const ShepherdFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_shepherd_free(ptr >>> 0, 1));

/**
 * @enum {0 | 1 | 2 | 3}
 */
export const AlertLevel = Object.freeze({
    Green: 0, "0": "Green",
    Yellow: 1, "1": "Yellow",
    Orange: 2, "2": "Orange",
    Red: 3, "3": "Red",
});

/**
 * Compression Dynamics Model for conflict potential calculation.
 *
 * Tracks actor "compression schemes" (worldviews) and computes
 * KL-divergence based conflict potential.
 */
export class CompressionModel {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        CompressionModelFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_compressionmodel_free(ptr, 0);
    }
    /**
     * Update an actor's scheme with a new observation.
     * @param {string} actor_id
     * @param {Float64Array} observation
     * @param {number} timestamp
     * @returns {boolean}
     */
    updateActor(actor_id, observation, timestamp) {
        const ptr0 = passStringToWasm0(actor_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passArrayF64ToWasm0(observation, wasm.__wbindgen_malloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.compressionmodel_updateActor(this.__wbg_ptr, ptr0, len0, ptr1, len1, timestamp);
        return ret !== 0;
    }
    /**
     * Get an actor's current entropy.
     * @param {string} actor_id
     * @returns {number | undefined}
     */
    actorEntropy(actor_id) {
        const ptr0 = passStringToWasm0(actor_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.compressionmodel_actorEntropy(this.__wbg_ptr, ptr0, len0);
        return ret[0] === 0 ? undefined : ret[1];
    }
    /**
     * Register a new actor with optional initial distribution.
     * @param {string} actor_id
     * @param {Float64Array | null} [distribution]
     */
    registerActor(actor_id, distribution) {
        const ptr0 = passStringToWasm0(actor_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(distribution) ? 0 : passArrayF64ToWasm0(distribution, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        wasm.compressionmodel_registerActor(this.__wbg_ptr, ptr0, len0, ptr1, len1);
    }
    /**
     * Set the learning rate for scheme updates.
     * @param {number} rate
     */
    setLearningRate(rate) {
        wasm.compressionmodel_setLearningRate(this.__wbg_ptr, rate);
    }
    /**
     * Compute conflict potential between two actors.
     * @param {string} actor_a
     * @param {string} actor_b
     * @returns {number | undefined}
     */
    conflictPotential(actor_a, actor_b) {
        const ptr0 = passStringToWasm0(actor_a, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(actor_b, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.compressionmodel_conflictPotential(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        return ret[0] === 0 ? undefined : ret[1];
    }
    /**
     * Get full conflict potential details as JSON.
     * @param {string} actor_a
     * @param {string} actor_b
     * @returns {any}
     */
    conflictPotentialDetails(actor_a, actor_b) {
        const ptr0 = passStringToWasm0(actor_a, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(actor_b, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.compressionmodel_conflictPotentialDetails(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        return ret;
    }
    /**
     * Create a new model with the specified number of categories.
     * @param {number} n_categories
     */
    constructor(n_categories) {
        const ret = wasm.compressionmodel_new(n_categories);
        this.__wbg_ptr = ret >>> 0;
        CompressionModelFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Get list of registered actors.
     * @returns {Array<any>}
     */
    actors() {
        const ret = wasm.compressionmodel_actors(this.__wbg_ptr);
        return ret;
    }
}
if (Symbol.dispose) CompressionModel.prototype[Symbol.dispose] = CompressionModel.prototype.free;

/**
 * Configuration for the variance inflection detector.
 */
export class DetectorConfig {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(DetectorConfig.prototype);
        obj.__wbg_ptr = ptr;
        DetectorConfigFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        DetectorConfigFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_detectorconfig_free(ptr, 0);
    }
    /**
     * @param {string} v
     */
    set kernel(v) {
        const ptr0 = passStringToWasm0(v, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.detectorconfig_set_kernel(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * @returns {number}
     */
    get window_size() {
        const ret = wasm.detectorconfig_window_size(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Create a conservative configuration.
     * @returns {DetectorConfig}
     */
    static conservative() {
        const ret = wasm.detectorconfig_conservative();
        return DetectorConfig.__wrap(ret);
    }
    /**
     * @param {number} v
     */
    set threshold(v) {
        wasm.detectorconfig_set_threshold(this.__wbg_ptr, v);
    }
    /**
     * @param {number} v
     */
    set window_size(v) {
        wasm.detectorconfig_set_window_size(this.__wbg_ptr, v);
    }
    /**
     * @returns {number}
     */
    get smoothing_window() {
        const ret = wasm.detectorconfig_smoothing_window(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get min_peak_distance() {
        const ret = wasm.detectorconfig_min_peak_distance(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} v
     */
    set smoothing_window(v) {
        wasm.detectorconfig_set_smoothing_window(this.__wbg_ptr, v);
    }
    /**
     * @param {number} v
     */
    set min_peak_distance(v) {
        wasm.detectorconfig_set_min_peak_distance(this.__wbg_ptr, v);
    }
    constructor() {
        const ret = wasm.detectorconfig_new();
        this.__wbg_ptr = ret >>> 0;
        DetectorConfigFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {string}
     */
    get kernel() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.detectorconfig_kernel(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Create a sensitive configuration.
     * @returns {DetectorConfig}
     */
    static sensitive() {
        const ret = wasm.detectorconfig_sensitive();
        return DetectorConfig.__wrap(ret);
    }
    /**
     * @returns {number}
     */
    get threshold() {
        const ret = wasm.detectorconfig_threshold(this.__wbg_ptr);
        return ret;
    }
}
if (Symbol.dispose) DetectorConfig.prototype[Symbol.dispose] = DetectorConfig.prototype.free;

/**
 * Variance Inflection Detector for phase transition detection.
 *
 * Detects phase transitions by finding peaks in the second derivative
 * of rolling variance.
 */
export class NucleationDetector {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NucleationDetector.prototype);
        obj.__wbg_ptr = ptr;
        NucleationDetectorFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NucleationDetectorFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_nucleationdetector_free(ptr, 0);
    }
    /**
     * Get confidence in the current assessment (0-1).
     * @returns {number}
     */
    confidence() {
        const ret = wasm.nucleationdetector_confidence(this.__wbg_ptr);
        return ret;
    }
    /**
     * Deserialize state from JSON string.
     * @param {string} json
     * @returns {NucleationDetector}
     */
    static deserialize(json) {
        const ptr0 = passStringToWasm0(json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.nucleationdetector_deserialize(ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return NucleationDetector.__wrap(ret[0]);
    }
    /**
     * Process multiple observations.
     * @param {Float64Array} values
     * @returns {Phase}
     */
    update_batch(values) {
        const ptr0 = passArrayF64ToWasm0(values, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.nucleationdetector_update_batch(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Get the current phase.
     * @returns {Phase}
     */
    currentPhase() {
        const ret = wasm.nucleationdetector_currentPhase(this.__wbg_ptr);
        return ret;
    }
    /**
     * Create a detector with default configuration.
     * @returns {NucleationDetector}
     */
    static with_defaults() {
        const ret = wasm.nucleationdetector_with_defaults();
        return NucleationDetector.__wrap(ret);
    }
    /**
     * Get the current rolling variance.
     * @returns {number}
     */
    currentVariance() {
        const ret = wasm.nucleationdetector_currentVariance(this.__wbg_ptr);
        return ret;
    }
    /**
     * Get the current inflection magnitude (z-score).
     * @returns {number}
     */
    inflectionMagnitude() {
        const ret = wasm.nucleationdetector_inflectionMagnitude(this.__wbg_ptr);
        return ret;
    }
    /**
     * Create a new detector with the given configuration.
     * @param {DetectorConfig} config
     */
    constructor(config) {
        _assertClass(config, DetectorConfig);
        const ret = wasm.nucleationdetector_new(config.__wbg_ptr);
        this.__wbg_ptr = ret >>> 0;
        NucleationDetectorFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Get the total number of observations processed.
     * @returns {number}
     */
    count() {
        const ret = wasm.nucleationdetector_count(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Reset the detector state.
     */
    reset() {
        wasm.nucleationdetector_reset(this.__wbg_ptr);
    }
    /**
     * Process a single observation.
     * @param {number} value
     * @returns {Phase}
     */
    update(value) {
        const ret = wasm.nucleationdetector_update(this.__wbg_ptr, value);
        return ret;
    }
    /**
     * Serialize state to JSON string.
     * @returns {string}
     */
    serialize() {
        let deferred2_0;
        let deferred2_1;
        try {
            const ret = wasm.nucleationdetector_serialize(this.__wbg_ptr);
            var ptr1 = ret[0];
            var len1 = ret[1];
            if (ret[3]) {
                ptr1 = 0; len1 = 0;
                throw takeFromExternrefTable0(ret[2]);
            }
            deferred2_0 = ptr1;
            deferred2_1 = len1;
            return getStringFromWasm0(ptr1, len1);
        } finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
}
if (Symbol.dispose) NucleationDetector.prototype[Symbol.dispose] = NucleationDetector.prototype.free;

/**
 * @enum {0 | 1 | 2 | 3}
 */
export const Phase = Object.freeze({
    Stable: 0, "0": "Stable",
    Approaching: 1, "1": "Approaching",
    Critical: 2, "2": "Critical",
    Transitioning: 3, "3": "Transitioning",
});

/**
 * Shepherd Dynamics: Unified early warning system.
 *
 * Combines compression dynamics with variance inflection detection
 * to identify "nucleation moments" before conflict escalation.
 */
export class Shepherd {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        ShepherdFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_shepherd_free(ptr, 0);
    }
    /**
     * Check a specific dyad for nucleation.
     * @param {string} actor_a
     * @param {string} actor_b
     * @param {number} timestamp
     * @returns {any}
     */
    checkDyad(actor_a, actor_b, timestamp) {
        const ptr0 = passStringToWasm0(actor_a, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(actor_b, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.shepherd_checkDyad(this.__wbg_ptr, ptr0, len0, ptr1, len1, timestamp);
        return ret;
    }
    /**
     * Get phi history for a dyad as Float64Array pairs [timestamp, phi, ...].
     * @param {string} actor_a
     * @param {string} actor_b
     * @returns {Float64Array}
     */
    phiHistory(actor_a, actor_b) {
        const ptr0 = passStringToWasm0(actor_a, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(actor_b, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.shepherd_phiHistory(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        return ret;
    }
    /**
     * Update an actor and check for nucleation alerts.
     * Returns array of alert objects.
     * @param {string} actor_id
     * @param {Float64Array} observation
     * @param {number} timestamp
     * @returns {Array<any>}
     */
    updateActor(actor_id, observation, timestamp) {
        const ptr0 = passStringToWasm0(actor_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passArrayF64ToWasm0(observation, wasm.__wbindgen_malloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.shepherd_updateActor(this.__wbg_ptr, ptr0, len0, ptr1, len1, timestamp);
        return ret;
    }
    /**
     * Register a new actor.
     * @param {string} actor_id
     * @param {Float64Array | null} [distribution]
     */
    registerActor(actor_id, distribution) {
        const ptr0 = passStringToWasm0(actor_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(distribution) ? 0 : passArrayF64ToWasm0(distribution, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        wasm.shepherd_registerActor(this.__wbg_ptr, ptr0, len0, ptr1, len1);
    }
    /**
     * Check all dyads for nucleation.
     * @param {number} timestamp
     * @returns {Array<any>}
     */
    checkAllDyads(timestamp) {
        const ret = wasm.shepherd_checkAllDyads(this.__wbg_ptr, timestamp);
        return ret;
    }
    /**
     * Get conflict potential between two actors.
     * @param {string} actor_a
     * @param {string} actor_b
     * @returns {number | undefined}
     */
    conflictPotential(actor_a, actor_b) {
        const ptr0 = passStringToWasm0(actor_a, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(actor_b, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.shepherd_conflictPotential(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        return ret[0] === 0 ? undefined : ret[1];
    }
    /**
     * Create a new Shepherd system.
     * @param {number} n_categories
     */
    constructor(n_categories) {
        const ret = wasm.shepherd_new(n_categories);
        this.__wbg_ptr = ret >>> 0;
        ShepherdFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Get list of registered actors.
     * @returns {Array<any>}
     */
    actors() {
        const ret = wasm.shepherd_actors(this.__wbg_ptr);
        return ret;
    }
}
if (Symbol.dispose) Shepherd.prototype[Symbol.dispose] = Shepherd.prototype.free;

/**
 * Compute Hellinger distance between two distributions.
 * @param {Float64Array} p
 * @param {Float64Array} q
 * @returns {number}
 */
export function hellingerDistance(p, q) {
    const ptr0 = passArrayF64ToWasm0(p, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF64ToWasm0(q, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.hellingerDistance(ptr0, len0, ptr1, len1);
    return ret;
}

/**
 * Compute Jensen-Shannon divergence between two distributions.
 * @param {Float64Array} p
 * @param {Float64Array} q
 * @returns {number}
 */
export function jensenShannonDivergence(p, q) {
    const ptr0 = passArrayF64ToWasm0(p, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF64ToWasm0(q, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.jensenShannonDivergence(ptr0, len0, ptr1, len1);
    return ret;
}

/**
 * Compute KL divergence between two distributions.
 * @param {Float64Array} p
 * @param {Float64Array} q
 * @returns {number}
 */
export function klDivergence(p, q) {
    const ptr0 = passArrayF64ToWasm0(p, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF64ToWasm0(q, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.klDivergence(ptr0, len0, ptr1, len1);
    return ret;
}

/**
 * Compute Shannon entropy of a distribution.
 * @param {Uint32Array} counts
 * @returns {number}
 */
export function shannonEntropy(counts) {
    const ptr0 = passArray32ToWasm0(counts, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.shannonEntropy(ptr0, len0);
    return ret;
}

/**
 * Get the library version.
 * @returns {string}
 */
export function version() {
    let deferred1_0;
    let deferred1_1;
    try {
        const ret = wasm.version();
        deferred1_0 = ret[0];
        deferred1_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
    }
}

const EXPECTED_RESPONSE_TYPES = new Set(['basic', 'cors', 'default']);

async function __wbg_load(module, imports) {
    if (typeof Response === 'function' && module instanceof Response) {
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            try {
                return await WebAssembly.instantiateStreaming(module, imports);
            } catch (e) {
                const validResponse = module.ok && EXPECTED_RESPONSE_TYPES.has(module.type);

                if (validResponse && module.headers.get('Content-Type') !== 'application/wasm') {
                    console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", e);

                } else {
                    throw e;
                }
            }
        }

        const bytes = await module.arrayBuffer();
        return await WebAssembly.instantiate(bytes, imports);
    } else {
        const instance = await WebAssembly.instantiate(module, imports);

        if (instance instanceof WebAssembly.Instance) {
            return { instance, module };
        } else {
            return instance;
        }
    }
}

function __wbg_get_imports() {
    const imports = {};
    imports.wbg = {};
    imports.wbg.__wbg___wbindgen_throw_dd24417ed36fc46e = function(arg0, arg1) {
        throw new Error(getStringFromWasm0(arg0, arg1));
    };
    imports.wbg.__wbg_new_1ba21ce319a06297 = function() {
        const ret = new Object();
        return ret;
    };
    imports.wbg.__wbg_new_25f239778d6112b9 = function() {
        const ret = new Array();
        return ret;
    };
    imports.wbg.__wbg_new_from_slice_9a48ef80d2a51f94 = function(arg0, arg1) {
        const ret = new Float64Array(getArrayF64FromWasm0(arg0, arg1));
        return ret;
    };
    imports.wbg.__wbg_new_with_length_806b9e5b8290af7c = function(arg0) {
        const ret = new Float64Array(arg0 >>> 0);
        return ret;
    };
    imports.wbg.__wbg_push_7d9be8f38fc13975 = function(arg0, arg1) {
        const ret = arg0.push(arg1);
        return ret;
    };
    imports.wbg.__wbg_set_781438a03c0c3c81 = function() { return handleError(function (arg0, arg1, arg2) {
        const ret = Reflect.set(arg0, arg1, arg2);
        return ret;
    }, arguments) };
    imports.wbg.__wbindgen_cast_2241b6af4c4b2941 = function(arg0, arg1) {
        // Cast intrinsic for `Ref(String) -> Externref`.
        const ret = getStringFromWasm0(arg0, arg1);
        return ret;
    };
    imports.wbg.__wbindgen_cast_d6cd19b81560fd6e = function(arg0) {
        // Cast intrinsic for `F64 -> Externref`.
        const ret = arg0;
        return ret;
    };
    imports.wbg.__wbindgen_init_externref_table = function() {
        const table = wasm.__wbindgen_externrefs;
        const offset = table.grow(4);
        table.set(0, undefined);
        table.set(offset + 0, undefined);
        table.set(offset + 1, null);
        table.set(offset + 2, true);
        table.set(offset + 3, false);
    };

    return imports;
}

function __wbg_finalize_init(instance, module) {
    wasm = instance.exports;
    __wbg_init.__wbindgen_wasm_module = module;
    cachedFloat64ArrayMemory0 = null;
    cachedUint32ArrayMemory0 = null;
    cachedUint8ArrayMemory0 = null;


    wasm.__wbindgen_start();
    return wasm;
}

function initSync(module) {
    if (wasm !== undefined) return wasm;


    if (typeof module !== 'undefined') {
        if (Object.getPrototypeOf(module) === Object.prototype) {
            ({module} = module)
        } else {
            console.warn('using deprecated parameters for `initSync()`; pass a single object instead')
        }
    }

    const imports = __wbg_get_imports();
    if (!(module instanceof WebAssembly.Module)) {
        module = new WebAssembly.Module(module);
    }
    const instance = new WebAssembly.Instance(module, imports);
    return __wbg_finalize_init(instance, module);
}

async function __wbg_init(module_or_path) {
    if (wasm !== undefined) return wasm;


    if (typeof module_or_path !== 'undefined') {
        if (Object.getPrototypeOf(module_or_path) === Object.prototype) {
            ({module_or_path} = module_or_path)
        } else {
            console.warn('using deprecated parameters for the initialization function; pass a single object instead')
        }
    }

    if (typeof module_or_path === 'undefined') {
        module_or_path = new URL('nucleation_bg.wasm', import.meta.url);
    }
    const imports = __wbg_get_imports();

    if (typeof module_or_path === 'string' || (typeof Request === 'function' && module_or_path instanceof Request) || (typeof URL === 'function' && module_or_path instanceof URL)) {
        module_or_path = fetch(module_or_path);
    }

    const { instance, module } = await __wbg_load(await module_or_path, imports);

    return __wbg_finalize_init(instance, module);
}

export { initSync };
export default __wbg_init;
